extensions/postgres/transactional.go

```diff
@@ -260,68 +261,64 @@ func (d dbTx) InsertLocalQueueMessage(ctx context.Context, row extensions.LocalQ
 	return effected == 1, err
 }
 
-func (d dbTx) InsertCustomTableErrorOnConflict(ctx context.Context, row extensions.CustomTableRowForInsert) error {
-	var cols []string
-	var vals []string
-	for k, v := range row.ColumnToValue {
-		cols = append(cols, k)
-		vals = append(vals, "'"+v+"'")
-	}
+func (d dbTx) InsertCustomTable(ctx context.Context, row extensions.CustomTableRow, writeConfigMode xcapi.WriteConflictMode) error {
+	var pkCols []string
+	var pkVals []string
+	var otherCols []string
+	var otherVals []string
 
-	_, err := d.tx.ExecContext(ctx,
-		`INSERT INTO `+row.TableName+` (`+row.PrimaryKey+`, `+strings.Join(cols, ", ")+`)
-    	VALUES ('`+row.PrimaryKeyValue+`', `+strings.Join(vals, ", ")+`)`)
-	return err
-}
+	for k, v := range row.PrimaryKeyColumnToValue {
+		pkCols = append(pkCols, k)
+		pkVals = append(pkVals, "'"+v+"'")
+	}
 
-func (d dbTx) InsertCustomTableIgnoreOnConflict(ctx context.Context, row extensions.CustomTableRowForInsert) error {
-	var cols []string
-	var vals []string
-	for k, v := range row.ColumnToValue {
-		cols = append(cols, k)
-		vals = append(vals, "'"+v+"'")
+	for k, v := range row.OtherColumnToValue {
+		otherCols = append(otherCols, k)
+		otherVals = append(otherVals, "'"+v+"'")
 	}
 
+	var onConflictClause string
+
+	switch writeConfigMode {
+	case xcapi.RETURN_ERROR_ON_CONFLICT:
+		onConflictClause = ""
+	case xcapi.IGNORE_CONFLICT:
+		onConflictClause = `ON CONFLICT DO NOTHING`
+	case xcapi.OVERRIDE_ON_CONFLICT:
+		var setClauses []string
+		for col, val := range row.OtherColumnToValue {
+			setClauses = append(setClauses, col+" = '"+val+"'")
+		}
+		updateClause := "UPDATE SET " + strings.Join(setClauses, ", ")
+
+		onConflictClause = `ON CONFLICT (` + strings.Join(pkCols, ", ") + `) DO ` + updateClause
+	default:
+		panic("unknown write mode " + string(writeConfigMode))
+	}
 	_, err := d.tx.ExecContext(ctx,
-		`INSERT INTO `+row.TableName+` (`+row.PrimaryKey+`, `+strings.Join(cols, ", ")+`)
-    	VALUES ('`+row.PrimaryKeyValue+`', `+strings.Join(vals, ", ")+`)
-		ON CONFLICT DO NOTHING`)
+		`INSERT INTO `+row.TableName+` (`+strings.Join(pkCols, ", ")+`, `+strings.Join(otherCols, ", ")+`)
+    	VALUES (`+strings.Join(pkVals, ", ")+`, `+strings.Join(otherVals, ", ")+`) `+onConflictClause)
 	return err
 }
 
-func (d dbTx) InsertCustomTableOverrideOnConflict(ctx context.Context, row extensions.CustomTableRowForInsert) error {
-	var cols []string
-	var vals []string
-	for k, v := range row.ColumnToValue {
-		cols = append(cols, k)
-		vals = append(vals, "'"+v+"'")
-	}
+func (d dbTx) UpsertCustomTableByPK(ctx context.Context, row extensions.CustomTableRow) error {
```

Comment:
for future optimization: we probably should support upsert multiple rows in a single query

you can add a todo here to do it later

Reply:
Created https://github.com/xcherryio/xcherry/issues/109